classes/boardmanager.php

```diff
@@ -0,0 +1,1152 @@
+     *
+     * @param int $id The board id
+     * @return void
+     */
+    public function delete_board(int $id) {
+        global $DB;
+        // Cards need to be read to identify files, assignees and discussions.
+        $cardids = $DB->get_fieldset_select('kanban_card', 'id', 'kanban_board = :id', ['id' => $id]);
+        $this->delete_cards($cardids);
+
+        $DB->delete_records('kanban_history', ['kanban_board' => $id]);
+        $DB->delete_records('kanban_column', ['kanban_board' => $id]);
+        $DB->delete_records('kanban_card', ['kanban_board' => $id]);
+        $DB->delete_records('kanban_board', ['id' => $id]);
+        // The rest of the elements is skipped in the update message.
+        $this->load_board($id);
+        $this->formatter->delete('board', ['id' => $id]);
+    }
+
+    /**
+     * Delete multiple cards and all attached data (discussions, assignees, files, calendar events).
+     *
+     * @param array $ids The card ids
+     * @param bool $updatecolumn Whether to update the column sequence (can be set to false, if column is going to be deleted)
+     * @return void
+     */
+    public function delete_cards(array $ids, bool $updatecolumn = true): void {
+        foreach ($ids as $id) {
+            $this->delete_card($id, $updatecolumn);
+        }
+    }
+
+    /**
+     * Delete a card and all attached data (discussions, assignees, files, calendar events).
+     *
+     * @param int $cardid Card id
+     * @param bool $updatecolumn Whether to update the column sequence (can be set to false, if column is going to be deleted)
+     * @return void
+     */
+    public function delete_card(int $cardid, bool $updatecolumn = true): void {
+        global $DB;
+        $fs = get_file_storage();
+        $DB->delete_records('kanban_discussion_comment', ['kanban_card' => $cardid]);
+        $DB->delete_records('kanban_assignee', ['kanban_card' => $cardid]);
+        $context = context_module::instance($this->cmid, IGNORE_MISSING);
+        $fs->delete_area_files($context->id, 'mod_kanban', 'attachments', $cardid);
+        $card = $this->get_card($cardid);
+        if ($updatecolumn) {
+            $column = $DB->get_record('kanban_column', ['id' => $card->kanban_column]);
+            $update = [
+                'id' => $column->id,
+                'timemodified' => time(),
+                'sequence' => helper::sequence_remove($column->sequence, $cardid),
+            ];
+            $DB->update_record('kanban_column', $update);
+            $this->formatter->put('columns', $update);
+            helper::update_cached_timestamp($card->kanban_board, constants::MOD_KANBAN_COLUMN);
+        }
+        $DB->delete_records('kanban_card', ['id' => $cardid]);
+        helper::remove_calendar_event($this->kanban, (object) ['id' => $cardid]);
+        $this->formatter->delete('cards', ['id' => $cardid]);
+        $this->write_history('deleted', constants::MOD_KANBAN_CARD, [], $card->kanban_column, $cardid);
```

Comment:
Well...if the card is deleted, doesn't history also disappear? Can anyone access the history after the card is being deleted? Is a cleanup somewhere happening?

Reply:
Not used now, history is cleaned when board is deleted or activity is reset.

Reply:
OK, as long as history is bound to a card, I would suggest to clean it up right here.